File: zZ/java/combination-sum-ii.java
/**
 * Combination Sum II
 * 
 * Given a collection of candidate numbers (C) and a target number (T), find all unique combinations in C where the candidate numbers sums to T.
 * 
 * Each number in C may only be used once in the combination.
 * 
 * Note:
 * 
 * All numbers (including target) will be positive integers.
 * Elements in a combination (a1, a2, … , ak) must be in non-descending order. (ie, a1 ≤ a2 ≤ … ≤ ak).
 * The solution set must not contain duplicate combinations.
 *  
 * 
 * For example, given candidate set 10,1,2,7,6,1,5 and target 8, 
 * A solution set is: 
 * [1, 7] 
 * [1, 2, 5] 
 * [2, 6] 
 * [1, 1, 6] 
 */

import java.util.Arrays;

public class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> path = new ArrayList<Integer>();
        Arrays.sort(candidates);
        combinationSumHelper(candidates, target, start, path, result);
        return result;
    }

    private void combinationSumHelper(int[] candidates, int target, int start, List<Integer> path, List<List<Integer>> result) {
        if (target == 0) {
            result.add(path);
        } else if (target > 0) {
            for (int i = start; i < candidates.length; ++i) {
                if (i > start && candidates[i] == candidates[i - 1]) {
                    continue;
                }
                path.add(candidates[i]);
                combinationSumHelper(candidates, target - candidates[i], i + 1, path, result);
                path.remove(path.size());
            }
        }
    }
}

File: data-structure-and-algorithm/typescript/leetcode/src/combination-sum-ii.ts
/*
Given a collection of candidate numbers (candidates) and a target number (target), find all unique combinations in candidates where the candidate numbers sum to target.

Each number in candidates may only be used once in the combination.

Note: The solution set must not contain duplicate combinations.



Example 1:

Input: candidates = [10,1,2,7,6,1,5], target = 8
Output:
[
[1,1,6],
[1,2,5],
[1,7],
[2,6]
]
Example 2:

Input: candidates = [2,5,2,1,2], target = 5
Output:
[
[1,2,2],
[5]
]


Constraints:

1 <= candidates.length <= 100
1 <= candidates[i] <= 50
1 <= target <= 30
*/

function combinationSumII(nums: number[], target: number): number[][] {
  const result: number[][] = [];
  nums.sort((a, b) => a - b);
  combinationSumIIDfs(target, 0, []);
  return result;

  function combinationSumIIDfs(target: number, start: number, path: number[]) {
    if (target === 0) {
      result.push([...path]);
      return;
    }
    for (let i = start; i < nums.length; ++i) {
      const diff = target - nums[i];
      if (diff >= 0 && (i === start || nums[i] !== nums[i - 1])) {
        path.push(nums[i]);
        combinationSumIIDfs(diff, i + 1, path);
        path.pop();
      }
    }
  }
}
